File: src/easy/Q0094.py
"""
94. Binary Tree Inorder Traversal
https://leetcode.com/problems/binary-tree-inorder-traversal/
"""
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# Recursive solution
# O(n) time, O(log n) space, where n is the number of nodes in the tree
def inorder_traversal_1(root: Optional[TreeNode]):
    if not root:
        return []
    return inorder_traversal_1(root.left) + [root.val] + inorder_traversal_1(root.right)


# Iterative solution
# O(n) time, O(n) space, where n is the number of nodes in the tree
def inorder_traversal_2(root: Optional[TreeNode]):
    result: List[int] = []
    stack: List[TreeNode] = []
    current = root
    while current or stack:
        while current:
            stack.append(current)
            current = current.left
        top = stack.pop()
        result.append(top.val)
        current = top.right
    return result

File: src/main/cpp/easy/Q0094.cpp
// 94. Binary Tree Inorder Traversal
// https://leetcode.com/problems/binary-tree-inorder-traversal/

#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
};

// Time: O(n), Space: O(n), where n is the number of nodes in the list.
auto inorder_traversal(TreeNode* root) -> vector<int> {
  vector<int> result;
  function<void(TreeNode*)> dfs = [&dfs, &result](TreeNode* root) {
    if (root == nullptr) {
      return;
    }
    dfs(root->left);
    result.push_back(root->val);
    dfs(root->right);
  };
  dfs(root);
  return result;
}
